appcues/src/main/java/com/appcues/Storage.kt

```diff
@@ -21,31 +25,85 @@ internal class Storage(
-        set(value) = sharedPreferences.edit().putString(Constants.UserId.rawVal, value).apply()
+        get() = _userId
+        set(value) = updateUserId(value)
+
+    private lateinit var _userId: String
+
+    private fun updateUserId(userId: String) {
+        _userId = userId
+        appcuesCoroutineScope.launch(Dispatchers.IO) {
+            sharedPreferences.edit().putString(Constants.UserId.rawVal, _userId).apply()
+        }
+    }
 
     var groupId: String?
-        get() = sharedPreferences.getString(Constants.GroupId.rawVal, null)
-        set(value) = sharedPreferences.edit().putString(Constants.GroupId.rawVal, value).apply()
+        get() = _groupId
+        set(value) = updateGroupId(value)
+
+    private var _groupId: String? = null
+
+    private fun updateGroupId(groupId: String?) {
+        _groupId = groupId
+        appcuesCoroutineScope.launch(Dispatchers.IO) {
+            sharedPreferences.edit().putString(Constants.GroupId.rawVal, _groupId).apply()
+        }
+    }
 
     var isAnonymous: Boolean
-        get() = sharedPreferences.getBoolean(Constants.IsAnonymous.rawVal, true)
-        set(value) = sharedPreferences.edit().putBoolean(Constants.IsAnonymous.rawVal, value).apply()
+        get() = _isAnonymous
+        set(value) = updateIsAnonymous(value)
+
+    private var _isAnonymous: Boolean = true
+
+    private fun updateIsAnonymous(isAnonymous: Boolean) {
+        _isAnonymous = isAnonymous
+        appcuesCoroutineScope.launch(Dispatchers.IO) {
+            sharedPreferences.edit().putBoolean(Constants.IsAnonymous.rawVal, _isAnonymous).apply()
+        }
+    }
 
     var lastContentShownAt: Date?
-        get() = sharedPreferences.getLong(Constants.LastContentShownAt.rawVal, 0).run { if (this > 0) Date(this) else null }
-        set(value) = when (value) {
-            null -> sharedPreferences.edit().remove(Constants.LastContentShownAt.rawVal).apply()
-            else -> sharedPreferences.edit().putLong(Constants.LastContentShownAt.rawVal, value.time).apply()
+        get() = _lastContentShownAt
+        set(value) = updateLastContentShownAt(value)
+
+    private var _lastContentShownAt: Date? = null
+
+    private fun updateLastContentShownAt(lastContentShownAt: Date?) {
+        _lastContentShownAt = lastContentShownAt
+        appcuesCoroutineScope.launch(Dispatchers.IO) {
+            _lastContentShownAt?.let {
+                sharedPreferences.edit().putLong(Constants.LastContentShownAt.rawVal, it.time).apply()
+            } ?: run {
+                sharedPreferences.edit().remove(Constants.LastContentShownAt.rawVal).apply()
+            }
         }
+    }
 
     init {
-        if (deviceId.isEmpty()) {
-            deviceId = UUID.randomUUID().toString()
+        appcuesCoroutineScope.launch {
+            if (sharedPreferences.getString(Constants.DeviceId.rawVal, String()).isNullOrEmpty()) {
+                _deviceId = UUID.randomUUID().toString()
+            }
```

Comment:
correct me if wrong, but this needs to set `deviceId` instead of `_deviceId` so it actually gets saved?  Otherwise each cold launch would generate a new device ID

Reply:
its not just that, but this is not actually working, I saw some crashing here. trying to look at it again.